File: learn/learn-algo/leetcode/630/go/630.go
package main

import (
	"container/heap"
	"fmt"
	"sort"
)

//func scheduleCourse(courses [][]int) int {
//	sort.Slice(courses, func(i, j int) bool {
//		return courses[i][1] < courses[j][1]
//	})
//
//	h := &Heap{}
//	total := 0 // 优先队列中所有课程的总时间
//	for _, course := range courses {
//		if t := course[0]; total+t <= course[1] {
//			total += t
//			heap.Push(h, t)
//		} else if h.Len() > 0 && t < h.IntSlice[0] {
//			total += t - h.IntSlice[0]
//			h.IntSlice[0] = t
//			heap.Fix(h, 0)
//		}
//	}
//	return h.Len()
//}

// ============实现heap===========
type Heap struct {
	sort.IntSlice
}

func (h Heap) Less(i, j int) bool {
	return h.IntSlice[i] > h.IntSlice[j]
}

func (h *Heap) Push(x interface{}) {
	h.IntSlice = append(h.IntSlice, x.(int))
	heap.Fix(h, 0)
}

func (h *Heap) Pop() interface{} {
	a := h.IntSlice
	x := a[len(a)-1]
	h.IntSlice = a[:len(a)-1]
	return x
}

func (h Heap) Top() int {
	return h.IntSlice[0]
}

func scheduleCourse(courses [][]int) int {
	sort.Slice(courses, func(i, j int) bool {
		return courses[i][1] < courses[j][1]
	})

	h := &Heap{}
	total := 0 // 优先队列中所有课程的总时间
	for _, course := range courses {
		if t := course[0]; total+t <= course[1] {
			total += t
			heap.Push(h, t)
		} else if h.Len() > 0 && t < h.IntSlice[0] {
			total += t - h.IntSlice[0]
			h.IntSlice[0] = t
			heap.Fix(h, 0)
		}
	}
	return h.Len()
}

func main() {
	var courses = [][]int{{5, 5}, {4, 6}, {2, 6}}
	res := scheduleCourse(courses)
	fmt.Println(res)
}

File: learn/learn-algo/leetcode/630/c++/630.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <numeric>
#include <math.h>
#include <queue>

using namespace std;

class Solution {
public:
    int scheduleCourse(vector<vector<int>>& courses) {
        sort(courses.begin(), courses.end(), [](auto a, auto b) { return a[1] < b[1];});
        priority_queue<int> heap;  // 默认是大顶堆
        int cnt = 0;
        for(auto & c : courses) {
            cnt += c[0];
            heap.push(c[0]);
            cout << "push";

            if(cnt > c[1]) {
                cnt -= heap.top();
                heap.pop();
            }
        }

        return heap.size();
    }
};


int main() {

//    Solution s;
//    vector<vector<int>> in = {{5, 5}, {4, 6}, {2, 6}};
//    auto res = s.scheduleCourse(in);
//    cout << res << endl;

    priority_queue<int> heap;  // 默认是大顶堆
    heap.push(5);
    heap.push(45);
    heap.push(10);
    heap.push(1);
    heap.push(100);
    heap.push(200);
    heap.push(500);
    heap.push(360);

    
    return 0;
}